README.md
service.subtitles.tusubtitulocom
==============================
tusubtitulo.com subtitle service plugin for XBMC/Kodi


changelog.txt
0.0.5.1
- changed to point to the new secure web pages: https

0.0.5
- adapt to new version of tusubtitulo.com

0.0.4
- adapt to new version of tusubtitulo.com

0.0.3
- BeautifulSoup, adapt to new version of tusubtitulo.com

0.0.2
- Code refactor, cleanup, optimization and tusubtitulo protection avoided (I think)

0.0.1
- Initial port from Subtitulos.es XBMC subtitles plugin by m0r3


resources/lib/utils.py
# -*- coding: utf-8 -*-
import urllib
import xbmc

languages = {
    u"Español (España)": ("Spanish", "es", "es", 1),
    u"Español (Latinoamérica)": ("Latino", "es", "es", 2),
    u"English": ("English", "en", "en", 2),
}

def log(module, msg):
    xbmc.log((u"### [%s] - %s" % (module,msg,)).encode('utf-8'), level=xbmc.LOGNOTICE)

def geturl(url):
    class AppURLopener(urllib.FancyURLopener):
        version = "Mozilla/5.0 (compatible, MSIE 11, Windows NT 6.3; Trident/7.0; rv:11.0) like Gecko"
        def __init__(self, *args):
            urllib.FancyURLopener.__init__(self, *args)
        def add_referrer(self, url=None):
            if url:
                urllib._urlopener.addheader('Referer', url)
    urllib._urlopener = AppURLopener()
    urllib._urlopener.add_referrer("https://www.tusubtitulo.com/")
    try:
        response = urllib._urlopener.open(url)
        content    = response.read()
    except:
        #log( __name__ ,"%s Failed to get url:%s" % ('test', url))
        content    = None
    return content


resources/lib/tusubtitulo.py
# -*- coding: utf-8 -*-

import re
from utils import languages, geturl, log
from BeautifulSoup import BeautifulSoup

main_url = "https://www.tusubtitulo.com/serie/"

def search_tvshow(tvshow, season, episode, languages):
    #log(__name__,"TVSHOW = %s" % (tvshow))
    code = None
    level = 0
    while code is None and level < 4:
        _tvshow, code = parsetvshow(tvshow, level)
        if code is None:
            level += 1
    #log(__name__, tvshow)
    #log(__name__, season)
    #log(__name__, episode)
    #log(__name__, languages)
    searchstring = re.sub(r'\s', '-', _tvshow + '/' + season + '/' + episode)
    url = main_url + searchstring.lower() + '/' + code
    #log( __name__ ,"URL = %s" % (url))
    return getallsubsforurl(url, languages)

def getcode(tvshow):
    #log(__name__, "CODE=%s" %(tvshow))
    reg_code = '<a href="/show/([0-9]+)">'+tvshow+'</a>'
    content = geturl("https://www.tusubtitulo.com/series.php")
    result = re.findall(reg_code, content, re.IGNORECASE | re.DOTALL | re.MULTILINE | re.UNICODE)
    return result[0] if len(result) > 0 else None

def parsetvshow(tvshow, level):
    parser = {0: lambda: tvshow,
               1: lambda: tvshow.replace('(', '').replace(')', '') if '(' in tvshow else None,
               2: lambda: re.sub(r'\s+\([^)]*\)', '', tvshow) if '(' in tvshow else None,
               3: lambda: tvshow.replace("'", '') if '\'' in tvshow else None}
    _tvshow = parser.get(level, lambda: None)()
    code = None
    if _tvshow is not None:
        code = getcode(_tvshow)
    #log(__name__, code)
    return (_tvshow, code)

def getallsubsforurl(url, langs):
    result = []
    content = geturl(url)
    soup = BeautifulSoup(content)
    for versions in soup.findAll("div", {"class": "ssdiv", "id": re.compile('version[0-9]+')}):
        version = versions.find("p", {"class": "title-sub"})
        #log(__name__, version)
        if version is None:
            continue
        version = version.text.split(',')[0]
        #log(__name__, version)
        for subtitulo in versions.findAll("ul", {"class":"sslist"}):
            idioma = subtitulo.findNext("li", {"class": "li-idioma"}).findNext("b").text
            #log(__name__, idioma)
            enlace = subtitulo.findNext("li", {"class": "rng download green"})
            if enlace is None:
                continue
            enlace = enlace.findNext("a", {"class": None})["href"]
            #log(__name__, enlace)
            _idioma = languages.get(idioma, ("Unknown", "-", "???", 3))
            if _idioma[1] not in langs:
                continue
            filename = "{0}".format(version.encode('utf-8'))
            result.append({'filename': filename, 'flag': _idioma[2],
                           'link': "https://www.tusubtitulo.com/"+enlace, 
                           'lang': _idioma[0],'order': 1 + _idioma[3], 'referer': url})
    return result

if __name__ == "__main__":
    subs = search_tvshow("game of thrones", "6", "1", "es,en")
    for sub in subs: print sub['server'], sub['link']


service.py
# -*- coding: utf-8 -*- 

import os
import sys
import xbmc
import urllib
import xbmcvfs
import xbmcaddon
import xbmcgui
import xbmcplugin
import re
import shutil
import unicodedata
import urllib2
import urlparse

__addon__ = xbmcaddon.Addon()
__author__     = __addon__.getAddonInfo('author')
__scriptid__   = __addon__.getAddonInfo('id')
__scriptname__ = __addon__.getAddonInfo('name')
__version__    = __addon__.getAddonInfo('version')
__language__   = __addon__.getLocalizedString

__cwd__        = xbmc.translatePath( __addon__.getAddonInfo('path') ).decode("utf-8")
__profile__    = xbmc.translatePath( __addon__.getAddonInfo('profile') ).decode("utf-8")
__resource__   = xbmc.translatePath( os.path.join( __cwd__, 'resources', 'lib' ) ).decode("utf-8")
__temp__       = xbmc.translatePath( os.path.join( __profile__, 'temp') ).decode("utf-8")

sys.path.append (__resource__)

from tusubtitulo import search_tvshow, log

""" Called when searching for subtitles from XBMC """
def Search(item):
    log(__name__, item)
    subs = search_tvshow(item['tvshow'], item['season'], item['episode'], item['2let_language'])
    for sub in subs:
        append_subtitle(sub)

def append_subtitle(item):
    listitem = xbmcgui.ListItem(label=item['lang'],  label2=item['filename'], thumbnailImage=item['flag'])
    url = "plugin://%s/?action=download&link=%s&filename=%s&referer=%s" % (__scriptid__, item['link'], item['filename'].decode('utf-8'), item['referer'])
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=url, listitem=listitem, isFolder=False)

def Download(link, filename, referer):
    subtitle_list = []
    if not link:
       return subtitle_list

    log(__name__, "Downloadlink %s" % link)

    class MyOpener(urllib.FancyURLopener):
        version = "Mozilla/5.0 (compatible, MSIE 11, Windows NT 6.3; Trident/7.0; rv:11.0) like Gecko"

    my_urlopener = MyOpener()
    my_urlopener.addheader('Referer', referer)
    log(__name__, referer)
    postparams = None

    log(__name__, "Fetching subtitles using url '%s' with referer header '%s' and post parameters '%s'" % (link, link, postparams))
    response = my_urlopener.open(link, postparams)
    local_tmp_file = os.path.join(__temp__, "sub.srt")

    if xbmcvfs.exists(__temp__):
        shutil.rmtree(__temp__)
    xbmcvfs.mkdirs(__temp__)
    try:
        log(__name__, "Saving subtitles to '%s'" % local_tmp_file)
        local_file_handle = open(local_tmp_file, "wb")
        local_file_handle.write(response.read())
        local_file_handle.close()

        subtitle_list.append(local_tmp_file)
        log(__name__, "=== returning subtitle file %s" % file)

    except:
        log(__name__, "Failed to save subtitle to %s" % local_tmp_file)

    return subtitle_list

def normalizeString(str):
  return unicodedata.normalize('NFKD', unicode(unicode(str, 'utf-8'))).encode('ascii','ignore')
 
def get_params():
  param = {}
  if len(sys.argv[2]) > 1:
      param = dict(urlparse.parse_qsl(sys.argv[2][1:]))
  return param


params = get_params()

if params['action'] == 'search':
  item = {}
  item['temp']               = False
  item['rar']                = False
  item['year']               = xbmc.getInfoLabel("VideoPlayer.Year")                           # Year
  item['season']             = str(xbmc.getInfoLabel("VideoPlayer.Season"))                    # Season
  item['episode']            = str(xbmc.getInfoLabel("VideoPlayer.Episode"))                   # Episode
  item['tvshow']             = normalizeString(xbmc.getInfoLabel("VideoPlayer.TVshowtitle"))   # Show
  item['title']              = normalizeString(xbmc.getInfoLabel("VideoPlayer.OriginalTitle")) # try to get original title
  item['file_original_path'] = urllib.unquote(xbmc.Player().getPlayingFile().decode('utf-8'))  # Full path of a playing file
  item['3let_language']      = []
  item['2let_language']      = []

  for lang in urllib.unquote(params['languages']).decode('utf-8').split(","):
    item['3let_language'].append(xbmc.convertLanguage(lang,xbmc.ISO_639_2))
    item['2let_language'].append(xbmc.convertLanguage(lang,xbmc.ISO_639_1))

  if item['title'] == "":
    item['title']  = normalizeString(xbmc.getInfoLabel("VideoPlayer.Title"))      # no original title, get just Title

  if item['episode'].lower().find("s") > -1:                                      # Check if season is "Special"
    item['season'] = "0"                                                          #
    item['episode'] = item['episode'][-1:]

  if ( item['file_original_path'].find("http") > -1 ):
    item['temp'] = True

  elif ( item['file_original_path'].find("rar://") > -1 ):
    item['rar']  = True
    item['file_original_path'] = os.path.dirname(item['file_original_path'][6:])

  elif ( item['file_original_path'].find("stack://") > -1 ):
    stackPath = item['file_original_path'].split(" , ")
    item['file_original_path'] = stackPath[0][8:]

  # required if tvshow is not indexed/recognized in library
  if item['tvshow'] == "":
    log(__name__, "item %s" % item)
    # replace dots with spaces in title
    titulo = re.sub(r'\.', ' ', item['title'])
    log(__name__, "title no dots: %s" % titulo)
    mo = re.search(r'(.*)[sS](\d+)[eE](\d+)', titulo) #S01E02 like
    if not mo:
      mo = re.search(r'(.*)(\d\d)[xX](\d+)', titulo) # old 10x02 style
    if not mo:
      mo = re.search(r'(.*)(\d)[xX](\d+)', titulo) # old 1x02 style
    if not mo:
      mo = re.search(r'(.*) (\d+)(\d\d)', titulo) # 102 style 
    # split title in tvshow, season and episode
    if mo:
      item['tvshow'] = mo.group(1)
      if item['tvshow'].endswith(' - '):
          item['tvshow'] = item['tvshow'][:-3]
      item['season'] = mo.group(2)
      item['episode'] = mo.group(3)
      log(__name__, "item %s" % item)
    else:
      log(__name__, "could not parse tvshow name and episode number")
  
  Search(item)  

elif params['action'] == 'download':
  log(__name__, params)
  ## we pickup all our arguments sent from def Search()
  subs = Download(params["link"],params["filename"],params['referer'])
  ## we can return more than one subtitle for multi CD versions, for now we are still working out how to handle that in XBMC core
  for sub in subs:
    listitem = xbmcgui.ListItem(label=sub)
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]),url=sub,listitem=listitem,isFolder=False)
 

xbmcplugin.endOfDirectory(int(sys.argv[1])) ## send end of directory to XBMC
